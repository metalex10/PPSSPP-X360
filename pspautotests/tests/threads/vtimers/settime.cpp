#include "shared.h"

inline void testSet(const char *title, SceUID vtimer, u64 t) {
	int result = sceKernelSetVTimerTime(vtimer, (SceKernelSysClock *)&t);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: OK %016x ", title, t);
	} else {
		checkpoint(NULL);
		schedf("%s: Failed (%08x) %016llx ", title, result, t);
	}
	schedfVTimer(vtimer);
}

inline void testSetWide(const char *title, SceUID vtimer, s64 t) {
	s64 updated = sceKernelSetVTimerTimeWide(vtimer, t);
	checkpoint(NULL);
	schedf("%s: %016llx ", title, updated);
	schedfVTimer(vtimer);
}

int countHits = 0;
SceUInt countHandler(SceUID uid, SceKernelSysClock *scheduled, SceKernelSysClock *actual, void *common) {
	++countHits;
	return 100000;
}

extern "C" int main(int argc, char *argv[]) {
	SceUID vtimer = sceKernelCreateVTimer("test", NULL);

	// Crashes.
	//sceKernelSetVTimerTime(vtimer, NULL);

	checkpointNext("Objects");
	testSet("  Normal", vtimer, 0x1337);
	testSet("  NULL", 0, 0x1337);
	testSet("  Invalid", 0xDEADBEEF, 0x1337);
	sceKernelDeleteVTimer(vtimer);
	testSet("  Deleted", vtimer, 0x1337);

	vtimer = sceKernelCreateVTimer("test", NULL);
	checkpointNext("Values");
	testSet("  1", vtimer, 1);
	testSet("  0x1234567890123", vtimer, 0x1234567890123ULL);
	testSet("  -1", vtimer, -1);
	testSet("  0", vtimer, 0);

	checkpointNext("Objects (wide)");
	testSetWide("  Normal", vtimer, 0);
	testSetWide("  NULL", 0, 0);
	testSetWide("  Invalid", 0xDEADBEEF, 0);
	sceKernelDeleteVTimer(vtimer);
	testSetWide("  Deleted", vtimer, 0);

	vtimer = sceKernelCreateVTimer("test", NULL);
	checkpointNext("Values (wide)");
	testSetWide("  1", vtimer, 1);
	testSetWide("  0x1234567890123", vtimer, 0x1234567890123ULL);
	testSetWide("  -1", vtimer, -1);
	testSetWide("  0", vtimer, 0);

	checkpointNext("While scheduled");
	sceKernelSetVTimerTimeWide(vtimer, 0);
	SceKernelSysClock t = {20000, 0};
	sceKernelSetVTimerHandler(vtimer, &t, &countHandler, (void *)0x00001337);
	sceKernelStartVTimer(vtimer);
	sceKernelDelayThread(1000);
	checkpoint("  After start: %d", countHits);
	sceKernelSetVTimerTimeWide(vtimer, 20000);
	checkpoint("  After set: %d", countHits);
	sceKernelDelayThread(1000);
	sceKernelStopVTimer(vtimer);
	checkpoint("  After set and delay: %d", countHits);

	SceKernelVTimerInfo info;
	info.size = sizeof(info);
	sceKernelReferVTimerStatus(vtimer, &info);
	checkpoint("%lld", *(u64 *)&info.schedule.low);

	sceKernelDeleteVTimer(vtimer);
	return 0;
}