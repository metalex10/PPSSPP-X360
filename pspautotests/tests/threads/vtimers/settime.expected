[x] Objects
[x]   Normal: OK 0000000000000000 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   NULL: Failed (800201be) 0000000000001337 VTimer: Invalid (800201be)
[x]   Invalid: Failed (800201be) 0000000000001337 VTimer: Failed (deadbeef)
[x]   Deleted: Failed (800201be) 0000000000001337 VTimer: Invalid (800201be)

[x] Values
[x]   1: OK 0000000000000000 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   0x1234567890123: OK 0000000000000001 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   -1: OK 0000000067890123 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   0: OK 00000000ffffffff VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=0,schedule=0)

[x] Objects (wide)
[x]   Normal: 0000000000000000 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=0,schedule=0)
[x]   NULL: ffffffffffffffff VTimer: Invalid (800201be)
[x]   Invalid: ffffffffffffffff VTimer: Failed (deadbeef)
[x]   Deleted: ffffffffffffffff VTimer: Invalid (800201be)

[x] Values (wide)
[x]   1: 0000000000000000 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   0x1234567890123: 0000000000000001 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   -1: 0001234567890123 VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=N,schedule=0)
[x]   0: ffffffffffffffff VTimer: (size=72,name=test,active=0,handler=0,common=00000000,base=0,current=0,schedule=0)

[x] While scheduled
[r]   After start: 0
[x]   After set: 0
[r]   After set and delay: 1
[x] 120000
